orchagent/orch.cpp

```diff
@@ -635,6 +635,138 @@ bool Orch::parseIndexRange(const string &input, sai_uint32_t &range_low, sai_uin
     return true;
 }
 
+/*
+ * generateBitMapFromIdsStr
+ *
+ * Generates the bit map representing the idsMap in string
+ * Args:
+ *      idsStr: The string representing the IDs.
+ *              Typically it's part of a key of BUFFER_QUEUE or BUFFER_PG, like "3-4" from "Ethernet0|3-4"
+ * Return:
+ *      idsMap: The bitmap of IDs. The LSB stands for ID 0.
+ *
+ * Example:
+ *      Input idsMap: 3-4
+ *      Return: 00001100b
+ */
+unsigned long Orch::generateBitMapFromIdsStr(const string &idsStr)
+{
+    sai_uint32_t lowerBound, upperBound;
+    unsigned long idsMap = 0;
+
+    if (!parseIndexRange(idsStr, lowerBound, upperBound))
+        return 0;
+
+    for (sai_uint32_t id = lowerBound; id <= upperBound; id ++)
+    {
+        idsMap |= (1 << id);
+    }
+
+    return idsMap;
+}
+
+/*
+ * generateIdListFromMap
+ *
+ * Parse the idsMap and generate a vector which contains slices representing bits in idsMap
+ * Args:
+ *     idsMap: The bitmap of IDs. The LSB stands for ID 0.
+ *     maxId: The maximum value of ID.
+ * Return:
+ *     A vector which contains slices representing bits in idsMap
+ *
+ * Example:
+ *     Input idsMap: 00100110b, maxId: 8
+ *     Return vector: ["1-2", "5"]
+ */
+set<string> Orch::generateIdListFromMap(unsigned long idsMap, sai_uint32_t maxId)
+{
+    unsigned long currentIdMask = 1;
+    bool started = false, needGenerateMap = false;
+    sai_uint32_t lower, upper;
+    set<string> extraIdsToReclaim;
+    for (sai_uint32_t id = 0; id <= maxId; id ++)
+    {
+        // currentIdMask represents the bit mask corresponding to id: (1<<id)
+        if (idsMap & currentIdMask)
```

Comment:
This logic does not seem to work if the bit map is 11100110 (basically if there is a range whose upper bound is maxId)

Reply:
Can you elaborate? I didn't find any issue and this is exactly how we use this function in reclaiming buffer.

Reply:
how does 9-15 get inserted into extraIdsToReclaim since needGenerateMap is set to False after you have inserted 7 and for 9-15,  L691 will be true and L709 will always be False. I am assuming the maxId in this case is 15 and after this we are done with the for loop. Did I miss something?

Reply:
In case the input is `111111010000000`, the maxId should be 16. (refer to the example: 0010011b, maxId: 8)
I believe that's the confusion, right?

Reply:
Adjusted the comment to make it clear.